runtime/src/gpu/cpu/gpu-cpu.c

```diff
@@ -0,0 +1,158 @@
+ * Other additional copyright holders may be indicated within.  *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifdef HAS_GPU_LOCALE
+
+#include "sys_basic.h"
+#include "chplrt.h"
+#include "chpl-mem.h"
+#include "chpl-mem-array.h"
+#include "chpl-gpu.h"
+#include "chpl-gpu-impl.h"
+#include "chpl-linefile-support.h"
+#include "chpl-tasks.h"
+#include "error.h"
+#include "chplcgfns.h"
+
+#include <assert.h>
+#include <stdbool.h>
+
+
+void chpl_gpu_impl_init(int* num_devices) {
+  CHPL_GPU_DEBUG("Initializing none GPU layer.\n");
+  *num_devices = 1;
+}
+
+static bool chpl_gpu_device_alloc = false;
+
+void chpl_gpu_impl_on_std_modules_finished_initializing(void) {
+  // The standard module has some memory that we allocate when we  are "on" a
+  // GPU sublocale when in fact we want to allocate it on the device. (As of
+  // the writing of this comment this is in `helpSetupLocaleGPU` in
+  // `LocaleModelHelpSetup`).
+  //
+  // Basically during the setup of the locale model we need to be "on" a given
+  // sublocale when we instantiate the object for it (the expectation is that
+  // the wide pointer for a sublocale appears to be on that sublocale),
+  // but in practice we don't actually want the data for the GPU sublocale
+  // object to be on the GPU).
+  //
+  // It's a bit of a hack but to handle this we start off setting
+  // `chpl_gpu_device_alloc` to false indicating that we shouldn't actually
+  // do any allocations on the device. Once the standard modules have finished
+  // loading this callback function
+  // (`chpl_gpu_impl_on_std_modules_finished_initializing`) gets called and we
+  // flip the flag.
+  chpl_gpu_device_alloc = true;
+}
+
+bool chpl_gpu_impl_is_device_ptr(const void* ptr) {
+  return false;  // this OK? maybe we want assertions to go through?
+}
+
+bool chpl_gpu_impl_is_host_ptr(const void* ptr) {
+  return true;
+}
+
+inline void chpl_gpu_impl_launch_kernel(int ln, int32_t fn,
+                                        const char* name,
+                                        int grd_dim_x,
+                                        int grd_dim_y,
+                                        int grd_dim_z,
+                                        int blk_dim_x,
+                                        int blk_dim_y,
+                                        int blk_dim_z,
+                                        int nargs, va_list args) {
```

Comment:
This should never be called right? Do we want to assert?

Reply:
I think this can be called (although not tested) if you were to use the kernel launch primitive instead of going through user-facing means.